ext/mbstring/libmbfl/mbfl/mbfilter.c

```diff
@@ -83,2575 +83,1531 @@
-	i = num;
-	while (--i >= 0) {
-		mbfl_identify_filter_cleanup(&flist[i]);
+	if (found_pos) {
+		return byte_offset_to_char_offset(haystack->encoding, haystack->val, haystack->len,
+			((unsigned char*)found_pos) - haystack->val);
 	}
 
-	efree((void *)flist);
-
-	return encoding;
+	return MBFL_ERROR_NOT_FOUND;
 }
 
 /*
- *  strlen
+ *  substr_count
  */
-static int
-filter_count_output(int c, void *data)
+size_t mbfl_substr_count(mbfl_string *haystack, mbfl_string *needle)
 {
-	(*(size_t *)data)++;
-	return c;
-}
+	ZEND_ASSERT(haystack->encoding == needle->encoding);
 
-size_t
-mbfl_strlen(const mbfl_string *string)
-{
-	size_t len, n, k;
-	unsigned char *p;
-	const mbfl_encoding *encoding = string->encoding;
+	if (mbfl_strlen(needle) == 0) {
+		return MBFL_ERROR_EMPTY;
+	}
 
-	len = 0;
-	if (encoding->flag & MBFL_ENCTYPE_SBCS) {
-		len = string->len;
-	} else if (encoding->flag & (MBFL_ENCTYPE_WCS2BE | MBFL_ENCTYPE_WCS2LE)) {
-		len = string->len/2;
-	} else if (encoding->flag & (MBFL_ENCTYPE_WCS4BE | MBFL_ENCTYPE_WCS4LE)) {
-		len = string->len/4;
-	} else if (encoding->mblen_table != NULL) {
-		const unsigned char *mbtab = encoding->mblen_table;
-		n = 0;
-		p = string->val;
-		k = string->len;
-		/* count */
-		if (p != NULL) {
-			while (n < k) {
-				unsigned m = mbtab[*p];
-				n += m;
-				p += m;
-				len++;
-			}
-		}
-	} else {
-		/* wchar filter */
-		mbfl_convert_filter *filter = mbfl_convert_filter_new(
-		  string->encoding,
-		  &mbfl_encoding_wchar,
-		  filter_count_output, 0, &len);
-		if (filter == NULL) {
-			return (size_t) -1;
-		}
-		/* count */
-		n = string->len;
-		p = string->val;
-		if (p != NULL) {
-			while (n > 0) {
-				(*filter->filter_function)(*p++, filter);
-				n--;
-			}
+	size_t result = 0;
+	unsigned char *search_pos = haystack->val, *found_pos, *e = search_pos + haystack -> len;
+
+	while (search_pos < e) {
+		found_pos = (unsigned char*)zend_memnstr(
+			(const char*)search_pos,
+			(const char*)needle->val, needle->len,
+			(const char*)haystack->val + haystack->len);
+		if (!found_pos) {
+			break;
 		}
-		mbfl_convert_filter_delete(filter);
+		result++;
+		search_pos = found_pos + needle->len;
 	}
 
-	return len;
+	return result;
 }
 
-
 /*
- *  strpos
+ *  substr
  */
-struct collector_strpos_data {
-	mbfl_convert_filter *next_filter;
-	mbfl_wchar_device needle;
-	size_t needle_len;
-	size_t start;
-	size_t output;
-	size_t found_pos;
-	size_t needle_pos;
-	size_t matched_pos;
-};
+mbfl_string *mbfl_substr(mbfl_string *string, mbfl_string *result, size_t from, size_t length)
+{
+	const mbfl_encoding *encoding = string->encoding;
+
+	mbfl_string_init_set(result, encoding);
 
-static int
-collector_strpos(int c, void* data)
+	size_t start = char_offset_to_byte_offset(encoding, string->val, string->len, from);
+	size_t end   = char_offset_to_byte_offset(encoding, string->val + start, string->len - start, length) + start;
+
+	result->val = ecalloc(end - start, sizeof(unsigned char));
```

Comment:
```suggestion
	result->val = emalloc(end - start);
```
`sizeof(unsigned char)` is always one, and there is no need to zero the buffer.

Even better would be:
```suggestion
	result->val = estrndup(start, end - start);
```
which will also properly null terminate the string (which we generally require in PHP for satefy reasons -- too many interactions with APIs that don't accept lengths).

Reply:
Thanks! I inherited the use of `sizeof(unsigned char)` from the previous implementation. Now that you mention it, it is clearly redundant. Thanks for pointing that out.

Null-terminating `result` here would be redundant; when strings are converted to `zend_string`, they are automatically null-terminated at that point.

Reply:
Hmm. But I see now that using `estrndup` is more concise, apart from any issue of whether the strings need to be null-terminated or not.